Add support for Zigbee binary sensors on ESP32 H2 and C6 (#5582)
OttoWinter/esphomedocs: src/content/docs/components/zigbee.mdx

---
description: "Instructions for setting up Zigbee component."
title: "Zigbee Component"
---

Zigbee is a low-power, mesh-based networking standard designed for IoT devices. Its low energy consumption makes
it well-suited for battery-powered smart-home devices, while its limited bandwidth is ideal for devices that exchange
only small amounts of data, such as switches, sensors, and motion detectors.

> Zigbee uses the same RF technology as Thread (IEEE 802.15.4) but defines also multiple application standards.

Reviewer: ```suggestion
Zigbee uses the same RF technology as Thread (IEEE 802.15.4) but also defines multiple application standards.
```

The `zigbee` component, however, supports only the Home Automation profile.
It allows exposing supported ESPHome components over a Zigbee network to Home Assistant via
**Zigbee2MQTT** or **ZHA**. Due to the limitations of the Zigbee protocol, only basic properties are exposed.
Additional properties must be configured manually in Home Assistant. Each ESPHome entity consumes one Zigbee endpoint.

> [!NOTE]
> You will need a Zigbee coordinator like **Zigbee2MQTT** or **ZHA**. Other commercial Zigbee hubs most likely don't
> support esphome components and might have issues with changing firmware.

> > When using Zigbee2MQTT make sure to use at least version 2.8.0. Otherwise names cannot have spaces in them and you need to setup at least 2 endpoints.

Reviewer: ```suggestion
> When using Zigbee2MQTT make sure to use at least version 2.8.0. Otherwise names cannot have spaces in them and you need to set up at least 2 endpoints.
```

## Supported Hardware

Zigbee is only supported on `nRF52` platforms and on `ESP32` with IEEE 802.15.4 connectivity (`ESP32-C5`, `ESP32-C6`, or `ESP32-H2` 
and in the future `ESP32-H4` and `ESP32-H21`). Both platforms support different features and have different limitations:
- On `nRF52` a maximum of 8 endpoints is supported.

> - `ESP32-H2` seems to work more reliable than `ESP32-C6`. There are reports of `ESP32-C6` boards that work only if placed next 

Reviewer: ```suggestion
- `ESP32-H2` seems to work more reliably than `ESP32-C6`. There are reports of `ESP32-C6` boards that work only if placed next 
```

to the coordinator. [Here is a list with known good boards.](https://github.com/luar123/zigbee_esphome/discussions/41#discussioncomment-14978980)
- On `ESP32` Zigbee can be used together with [wifi](/components/wifi) with some limitations:
  - Wifi station mode + Zigbee end device is working fine.

>   - Wifi station mode + Zigbee router leads to a instable Zigbee network. Still can be used if wifi is only enabled for short times (e.g. OTA)

Reviewer: ```suggestion
  - Wifi station mode + Zigbee router leads to an unstable Zigbee network. Still can be used if wifi is only enabled for short times (e.g. OTA)
```

>   - Wifi access point + Zigbee end device could lead to a unstable access point with increased package loss. If the AP is only used as fallback there should be no issues.

Reviewer: ```suggestion
  - Wifi access point + Zigbee end device could lead to an unstable access point with increased packet loss. If the AP is only used as fallback there should be no issues.
```

it well-suited for battery-powered smart-home devices, while its limited bandwidth is ideal for devices that exchange
only small amounts of data, such as switches, sensors, and motion detectors.

Zigbee uses the same RF technology as Thread (IEEE 802.15.4) but also defines multiple application standards.
The `zigbee` component, however, supports only the Home Automation profile.
It allows exposing supported ESPHome components over a Zigbee network to Home Assistant via
**Zigbee2MQTT** or **ZHA**. Due to the limitations of the Zigbee protocol, only basic properties are exposed.
Additional properties must be configured manually in Home Assistant. Each ESPHome entity consumes one Zigbee endpoint.

> [!NOTE]
> You will need a Zigbee coordinator like **Zigbee2MQTT** or **ZHA**. Other commercial Zigbee hubs most likely do not
> support ESPHome components and might have issues with changing firmware.
> When using Zigbee2MQTT, make sure to use at least version 2.8.0. Otherwise, names cannot contain spaces.
> You also need to set up at least 2 endpoints in this case.

## Supported Hardware

> Zigbee is only supported on `nRF52` platforms and on `ESP32` with IEEE 802.15.4 connectivity (`ESP32-C5`, `ESP32-C6`, or `ESP32-H2` 

Author: This is not correct

and in the future `ESP32-H4` and `ESP32-H21`). Both platforms support different features and have different limitations:
- On `nRF52` a maximum of 8 endpoints is supported.
- `ESP32-H2` seems to work more reliably than `ESP32-C6`. There are reports of `ESP32-C6` boards that work only if placed next 

> to the coordinator. [Here is a list with known good boards.](https://github.com/luar123/zigbee_esphome/discussions/41#discussioncomment-14978980)

Reviewer: I would remove the link to your own gh.

  on_join:
    then:
      - logger.log: "Joined network"
```

## Configuration variables

> - **model** (*Optional*, string): Model Identifier of the Zigbee device. Defauts to the ESPHome node name.

Reviewer: ```suggestion
- **model** (*Optional*, string): Model Identifier of the Zigbee device. Defaults to the ESPHome node name.
```

- **id** (*Optional*, [ID](/guides/configuration-types#id)): The ID to use for this `zigbee` component.

> Only on `nrf52`:

Reviewer: ```suggestion
Only on `nRF52`:
```
Same on line 108

>   on_join:

Reviewer: `on_join` is nRF52 only but the example uses it without noting that. Either split the example or add a comment.

Author: done

> [!WARNING]
> Overusing `random` may exhaust memory in the Zigbee coordinator by creating many "ghost" devices.

Only on `ESP32`:

- **router** (*Optional*, boolean): If true the Zigbee role will be `router` instead of `end device` and it will

>  forward Zigbee packets from other devices. Don't use this for battery powered devices. Defaults to `false`.

Reviewer: ```suggestion
- **router** (*Optional*, boolean): If true the Zigbee role will be `router` instead of `end device` and it will
  forward Zigbee packets from other devices. Don't use this for battery-powered devices. Defaults to `false`.
```

    - zigbee.factory_reset
```

## Zigbee Component Base Configuration

All components in ESPHome that do some sort of communication through
Zigbee can have some overrides for specific options.

### Configuration variables

- **name** (**Required**, string): The name for the component. This is exposed as the Zigbee description attribute.
- **internal** (*Optional*, boolean): Mark this component as internal. Internal components will
  not be exposed over Zigbee. Only specifying an `id` without a `name` will implicitly set this to true.
  Use this if you run out of Zigbee endpoints.

Only on `ESP32`:

> - **report** (**Optional**, enum): Report the state. One of `coordinator`, `enable`, `force`. `coordinator` uses the configuration from

Reviewer: ```suggestion
- **report** (*Optional*, enum): Report the state. One of `coordinator`, `enable`, `force`. `coordinator` uses the configuration from
```
Other entries use `*Optional*` (italic), this one is bold.

  the coordinator, `enable` activates reporting with default settings and uses the configuration from the coordinator if possible.

>   `force` ignores the coordinator configuration and reports every change. On `nRF52` always the coordinator configuration is used.

Reviewer: ```suggestion
  `force` ignores the coordinator configuration and reports every change. On `nRF52` the coordinator configuration is always used.
```

  Defaults to `coordinator`.

> [!NOTE]
> ZHA sets the minimum reporting interval to 30 seconds for most sensor devices. If you need faster responses set `report` to `force`.

> ### Supported Components

Reviewer: Why is `Supported Components` nested under `Zigbee Component Base Configuration`? Seems like it should be its own top-level `##` section

The following components are exposed over Zigbee if they have a name and if they are not marked as internal:

- [Binary Sensor](/components/binary_sensor#config-binary_sensor): only **state** and **name** are exposed over Zigbee.

Only on `nRF52`:

- [Sensor](/components/sensor#config-sensor): **state**, **name** and **unit_of_measurement** are exposed
  over Zigbee. If no unit is given or if it is not supported it will fall back to unitless.

- [Switch](/components/switch#config-switch): only **state** and **name** are exposed over Zigbee.

- [Number](/components/number#config-number): **state**, **name**, **unit_of_measurement**, **min_value**, **max_value**, and **step** are exposed
  over Zigbee. If no unit is given or if it is not supported it will fall back to unitless.

> ## Component Implementation Details

Reviewer: Should note sensor/switch/number sections below are nRF52 only since those arent exposed on ESP32 yet

Author: done

### Sensor

[Sensors](/components/sensor#config-sensor) are implemented as analog input clusters with description, engineering_units,
and present_value attributes.

#### Additional configuration variables

- **unit_of_measurement** (*Optional*, string): Manually set or overwrite the unit. By default
  the unit of the component is used and if the component does not set a unit, values are unitless.
  Only a limited set of units is supported. Unsupported units will revert to unitless.
  This is exposed as the Zigbee engineering units attribute.

### Switch

[Switches](/components/switch#config-switch) are implemented as binary output clusters with description and present_value attributes.

### Number

[Numbers](/components/number#config-number) are implemented as analog output clusters with description, engineering_units,
and present_value attributes.

#### Additional configuration variables

- **unit_of_measurement** (*Optional*, string): Manually set the unit. By default, values are unitless.
  Only a limited set of units is supported. Unsupported units will revert to unitless.

>   This is exposed as the Zigbee engineering units attribute.

Reviewer: Wording here differs from the Sensor `unit_of_measurement` entry above (line 158). Should match.

Author: This is because numbers don't have a default unit compared to most sensor components that have one that you still can overwrite.

## Supported Components

The following components are exposed over Zigbee if they have a name and if they are not marked as internal:

- [Binary Sensor](/components/binary_sensor#config-binary_sensor): only **state** and **name** are exposed over Zigbee.

Only on `nrf52`:

- [Sensor](/components/sensor#config-sensor): **state**, **name** and **unit_of_measurement** are exposed
over Zigbee. If no unit is given or if it is not supported it will fall back to unitless.

- [Switch](/components/switch#config-switch): only **state** and **name** are exposed over Zigbee.

- [Number](/components/number#config-number): **state**, **name**, **unit_of_measurement**, **min_value**, **max_value**, and **step** are exposed
over Zigbee. If no unit is given or if it is not supported it will fall back to unitless.

> ## Zigbee Component Base Configuration

Reviewer: This section describes options that apply to all components (name/internal/report). Feels like it belongs *before* the "Supported Components" list, or merge it into a single section — current ordering makes the reader backtrack

All components in ESPHome that do some sort of communication through
Zigbee can have some overrides for specific options.

### Configuration variables

- **name** (**Required**, string): The name for the component. This is exposed as the Zigbee description attribute.
- **internal** (*Optional*, boolean): Mark this component as internal. Internal components will
  not be exposed over Zigbee. Only specifying an `id` without a `name` will implicitly set this to true.
  Use this if you run out of Zigbee endpoints.

Only on `ESP32`:

- **report** (**Optional**, enum): Report the state. One of `yes`, `no`, `force`. `yes` activates reporting with default settings and uses
the configuration from the coordinator if possible. `force` ignores the the coordinator settings and reports

> every change. On `nrf52` always the coordinator settings are used. Defaults to `yes`.

Reviewer: ```suggestion
- **report** (**Optional**, enum): Report the state. One of `yes`, `no`, `force`. `yes` activates reporting with default settings and uses
the configuration from the coordinator if possible. `force` ignores the coordinator settings and reports
every change. On `nRF52` always the coordinator settings are used. Defaults to `yes`.
```
`yes`/`no` as enum values is fragile in YAML (they can parse as booleans). Worth considering `always`/`never`/`force` or similar
